js-peer/src/lib/direct-message.ts

```diff
@@ -0,0 +1,180 @@
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class DirectMessage extends TypedEventEmitter<DirectMessageEvents> implements Startable {
+  readonly [serviceDependencies]: string[] = [
+    '@libp2p/identify',
+    '@libp2p/connection-encryption',
+    '@libp2p/transport',
+    '@libp2p/stream-multiplexing',
+  ]
+
+  readonly [serviceCapabilities]: string[] = [
+    '@universal-connectivity/direct-message'
+  ]
+
+  private topologyId?: string
+  private readonly components: DirectMessageComponents
+  private dmPeers: Set<string> = new Set()
+
+  constructor(components: DirectMessageComponents) {
+    super()
+    this.components = components
+  }
+
+  async start(): Promise<void> {
+    this.topologyId = await this.components.registrar.register(DIRECT_MESSAGE_PROTOCOL, {
+      onConnect: this.handleConnect.bind(this),
+      onDisconnect: this.handleDisconnect.bind(this)
+    })
+  }
+
+  async afterStart(): Promise<void> {
+    await this.components.registrar.handle(
+      DIRECT_MESSAGE_PROTOCOL,
+      async ({ stream, connection }) => {
+        await this.receive(stream, connection)
+      },
+    )
+  }
+
+  stop(): void {
+    if (this.topologyId != null) {
+      this.components.registrar.unregister(this.topologyId)
+    }
+  }
+
+  private handleConnect(peerId: PeerId): void {
+    this.dmPeers.add(peerId.toString())
+  }
+
+  private handleDisconnect(peerId: PeerId): void {
+    this.dmPeers.delete(peerId.toString())
+  }
+
+  isDMPeer(peerId: PeerId): boolean {
+    return this.dmPeers.has(peerId.toString())
+  }
+
+  async send(peerId: PeerId, message: string): Promise<boolean> {
+    if (!message) {
+      throw new Error(ERRORS.EMPTY_MESSAGE)
+    }
+
+    let stream: Stream | undefined
+
+    try {
+      const conn = await this.components.connectionManager.openConnection(peerId, { signal: AbortSignal.timeout(5000) })
+      if (!conn) {
+        throw new Error(ERRORS.NO_CONNECTION)
+      }
+
+      // Single protocols can skip full negotiation
+      const stream = await conn.newStream(DIRECT_MESSAGE_PROTOCOL, { negotiateFully: false })
+
+      if (!stream) {
+        throw new Error(ERRORS.NO_STREAM)
+      }
+
+      const datastream = pbStream(stream)
+
+      const req: dm.DirectMessageRequest = {
+        content: message,
+        type: MIME_TEXT_PLAIN,
+        metadata: {
+          clientVersion: dmClientVersion,
+          timestamp: BigInt(Date.now()),
+        },
+      }
+
+      await datastream.write(req, dm.DirectMessageRequest)
+
+      const res = await datastream.read(dm.DirectMessageResponse)
+
+      if (!res) {
+        throw new Error(ERRORS.NO_RESPONSE)
+      }
+
+      if (!res.metadata) {
+        throw new Error(ERRORS.NO_METADATA)
+      }
+
+      if (res.status !== dm.Status.OK) {
+        throw new Error(ERRORS.STATUS_NOT_OK(res.status))
+      }
+    } catch (e: any) {
+      stream?.abort(e)
+      throw e
+    } finally {
+      await stream?.close()
```

Comment:
It's a good idea to pass a signal when closing a stream so you don't (potentially) wait for a long time.  If closing the stream throws for whatever reason, it's also a good idea to abort the stream so it's reset for the remote and all resources are cleaned up:

```suggestion
      try {
        await stream?.close({
          signal: AbortSignal.timeout(5000)
        })      
      } catch (err: any) {
        stream?.abort(err)
        throw err
      }
```